src/client/core/aad/auth-provider.ts

```diff
@@ -63,45 +79,80 @@ export default class AuthProvider {
                 );
                 code = await authCodeCallback(url, true);
-            } catch (e) {
+            } catch (silentAuthExcpetion) {
                 url = await client.getAuthCodeUrl(authRequest);
                 code = await authCodeCallback(url);
             }
 
             const result: AuthorizationResult =
                 await client.acquireTokenByCode({ ...authRequest, code });
-            if (result) {
-                this._accounts[tenantId] = result.account;
-            }
             return result;
         }
     }
 
-    public logout(): void {
-        this._removeAccount();
+    public async logout(): Promise<void> {
+        this._logoutPromise = new Promise(resolve => {
+            this._removeAccounts().then(resolve);
+        });
     }
 
     protected _getClient(tenantId: string): ClientApplication {
         if (tenantId in this._clients) {
             return this._clients[tenantId];
         }
-        return this._clients[tenantId] = new PublicClientApplication({
+        const client = new PublicClientApplication({
             auth: {
                 clientId: this.config.clientId,
                 authority:
                     `${this.app.properties.azureEnvironment.aadUrl}${tenantId}/`
+            },
+            cache: {
+                cachePlugin: this._cachePlugin
             }
         });
+        if (!this._primaryClient) {
+            this._primaryClient = client;
+        }
+        this._clients[tenantId] = client;
+        return client;
     }
 
-    private async _removeAccount(): Promise<void> {
-        for (const tenantId in this._clients) {
-            if (this._accounts[tenantId]) {
-                const cache = this._clients[tenantId].getTokenCache();
-                cache.removeAccount(this._accounts[tenantId]);
+    private async _removeAccounts(): Promise<void> {
+        const cache = this._primaryClient?.getTokenCache();
+        const accounts = await cache.getAllAccounts();
+        for (const account of accounts) {
+            await cache.removeAccount(account);
+        }
+        this._accounts = {};
+        this._clients = {};
+        this._primaryClient = undefined;
+    }
+
+    private async _getAccount(tenantId: string): Promise<AccountInfo> {
+        if (tenantId in this._accounts) {
+            return this._accounts[tenantId];
+        }
+        const cache = this._clients[tenantId].getTokenCache();
+        const accounts: AccountInfo[] = await cache.getAllAccounts();
+        let homeAccountId = null;
+        for (const account of accounts) {
+            if (account.tenantId === tenantId) {
+                return this._accounts[tenantId] = account;
+            } else if (!homeAccountId) {
+                homeAccountId = account.homeAccountId;
             }
-            delete this._accounts[tenantId];
         }
+
+        /* SPECIAL CASE: If the tenant is one of the tenant placeholders (e.g.,
```

Comment:
Maybe another good place to put some logging?

Reply:
Does this really need to be logged? In 99% of the cases, users will be authenticated against a tenant placeholder, so this is a standard code path. It's only a special case in the sense that we're returning the account for the tenant placeholder using its home account ID as opposed to its resolved ID.

Reply:
Good point - I'm still a little concerned about edge cases here. The return type should also be `Promise<AccountInfo | null>` since we're returning null at the end, and we should make sure other code is handling that null. (If that case will never happen, then it should throw an exception instead).

Reply:
Update: See my other comment on usage of this method. I think we should just throw an error if there's no better graceful way to handle a missing account.